### PR TITLE
Support DoRA (use_dora=True) in the safetensors LoRA merge

### PR DESCRIPTION
**Pull request by danielhanchen**
### Summary

DoRA (`use_dora=True`) adapters could not be merged through the safetensors path. `create_lora_statistics` did not capture the `lora_magnitude_vector`, so the key-consistency check raised `Unsloth: Extracted keys = {...lora_magnitude_vector.default.weight} do not match!`, and even past that the merge math applied only `W0 + alpha*BA`, dropping the magnitude.

This adds DoRA support for the dense merge path (attention / MLP / any dense linear, including the language and vision/audio towers of multimodal models):

- `LoraStats` gains a `magnitude` field; `create_lora_statistics` captures `lora_magnitude_vector.default`.
- `assert_same_keys` ignores the magnitude tensor (folded into the weight, omitted from the merged checkpoint), like `lora_A` / `lora_B`.
- `_merge_lora` folds it in as `(m / ||W0 + alpha*BA||_row) * (W0 + alpha*BA)`, i.e. one L2 norm per output row over the input dim, matching PEFT's DoRA merge.

MoE-expert DoRA is explicitly refused with a clear error (the per-expert / fused expert helpers fold only the LoRA delta, not the magnitude), so a DoRA MoE merge fails loud instead of silently dropping the magnitude.

### Test

`tests/test_dora_merge.py`:
- dense DoRA merge equals PEFT's own `merge_and_unload` (atol 1e-4) and actually changes the weight;
- plain LoRA (`use_dora=False`) is unchanged (`magnitude is None`);
- DoRA on a MoE expert raises.

Full merge e2e suite (dense + MoE fused/per-expert + resized) stays green (65 passed, 6 skipped).

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/saving_utils.py [hunk 1]
```diff
@@ -344,13 +358,15 @@ def assert_same_keys(model, new_state_dict):
...
-        # state_dict intentionally omits lora_A / lora_B weights.
+        # state_dict intentionally omits lora_A / lora_B / DoRA magnitude weights
+        # (the magnitude is folded into the merged weight in _merge_lora).
         return (
             "modules_to_save" in key
             or "original_module" in key
             or ".lora_A" in key
             or ".lora_B" in key
             or ".lora_embedding" in key
+            or ".lora_magnitude_vector" in key
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/saving_utils.py:369]: Confirmed and fixed. PEFT stores the embedding DoRA delta as lora_embedding_A/lora_embedding_B with a separate magnitude, which this merge does not read, so the magnitude was captured with no mergeable delta and silently dropped (and assert_same_keys now ignores magnitude keys). create_lora_statistics now fails loud when a magnitude is present but lora_A/lora_B are None, matching the MoE DoRA refuse.

File: unsloth_zoo/saving_utils.py [hunk 2]
```diff
@@ -1179,10 +1201,24 @@ def _detect_moe_lora_layout(lora_A, lora_B, num_experts, out_dim, in_dim, lora_m
     return "unknown", r
 
 
+def _refuse_dora_on_moe(lora_stats):
+    """DoRA on MoE experts is not yet supported: the expert merge helpers fold only the LoRA
+    delta, not the DoRA magnitude (the dense path handles it in _merge_lora). Fail loud rather
+    than emit a checkpoint with the magnitude silently dropped."""
+    if getattr(lora_stats, "magnitude", None) is not None:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/saving_utils.py:1228]: Confirmed and fixed. The MXFP4 packed-expert rewrite merges via _merge_lora directly rather than the dense _merge_moe_*_expert helpers, so it never reached _refuse_dora_on_moe and a use_dora adapter crashed with an opaque 3D expand error. Added the refuse to the packed-expert branch.

File: unsloth_zoo/saving_utils.py [hunk 3]
```diff
@@ -243,6 +243,19 @@ def _merge_lora(W, lora_stats, name, use_dequant_base = False):
...
+    magnitude = getattr(lora_stats, "magnitude", None)
+    if magnitude is not None:
+        magnitude = magnitude.to(device, dtype = torch.float32, non_blocking = True).reshape(-1)
+        if magnitude.shape[0] != W.shape[0]:
+            raise ValueError(
+                f"Unsloth: DoRA magnitude for `{name}` has {magnitude.shape[0]} entries but the "
+                f"merged weight has {W.shape[0]} output rows."
+            )
+        weight_norm = torch.linalg.norm(W, dim = 1).clamp_min(1e-9)
+        W = (magnitude / weight_norm).unsqueeze(1) * W
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/saving_utils.py:258]: The DoRA norm is never computed on a raw quantized/packed base here: merged_16bit uses the downloaded 16-bit base (nf4/fp4 to merged_16bit is refused earlier), the gated-arch path substitutes dequant(W4) before the norm via use_dequant_base, and the in-memory/mxfp4 paths dequantize first. The merged per-row norm equals the magnitude m exactly in every case; only the row direction differs by the O(quant-error) term, the same approximation already accepted for plain LoRA. PEFT's dequantize-before-norm changes only that direction, not correctness.

**Comment by danielhanchen** [comment]
@codex review
